go/common/crypto/signature/signer_test.go

```diff
@@ -0,0 +1,55 @@
+package signature
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestContext(t *testing.T) {
+	require := require.New(t)
+
+	// Make sure a context can be registered.
+	var ctx Context
+	require.NotPanics(func() { ctx = NewContext("test: dummy context 1") })
+
+	// Make sure we panic if the same context is registered twice.
+	require.Panics(func() { NewContext("test: dummy context 1") })
+	// Even with different options.
+	require.Panics(func() { NewContext("test: dummy context 1", WithChainSeparation()) })
+
+	// Make sure we panic if context is too long.
+	require.Panics(func() { NewContext(strings.Repeat("a", 500)) })
+
+	// Make sure we panic if context includes the chain context separator.
+	require.Panics(func() { NewContext("test: dummy context 1 for chain blah") })
+
+	// Make sure a context with chain separation can be registered.
+	var chainCtx Context
+	require.NotPanics(func() { chainCtx = NewContext("test: dummy context 2", WithChainSeparation()) })
+
+	// Make sure we cannot use an unregistered context.
+	unregCtx := Context("test: unregistered")
+	_, err := PrepareSignerMessage(unregCtx, []byte("message"))
+	require.Error(err, "PrepareSignerMessage should fail with unregistered context")
+
+	// Should work with registered context.
+	_, err = PrepareSignerMessage(ctx, []byte("message"))
+	require.NoError(err, "PrepareSignerMessage should work")
+
+	// Should fail with context that requires chain separation as no
+	// chain context has been set.
+	_, err = PrepareSignerMessage(chainCtx, []byte("message"))
+	require.Error(err, "PrepareSignerMessage should fail without chain context")
+
+	// Make sure we can set a chain context.
+	require.NotPanics(func() { SetChainContext("test: oasis-core tests") })
+	// Make sure we can set the same chain context again.
+	require.NotPanics(func() { SetChainContext("test: oasis-core tests") })
+	// Make sure we can't modify a set chain context.
+	require.Panics(func() { SetChainContext("test: context change") })
+
+	_, err = PrepareSignerMessage(chainCtx, []byte("message"))
+	require.NoError(err, "PrepareSignerMessage should work with chain context")
```

Comment:
Should probably test that the signer messages are different with and without contexts.

Reply:
Good idea, added.